added phpunit tests and found a bug

Zae opened the pull request:
First I wanted to change the library to make it possible to call the methods non-statically, after doing so I could not figure out why phpspec was not throwing any errors as it should, so I rewrote all the specs in phpunit.

Then I found out PHP allows you to call static methods non-statically so all my work was for nothing, or so I thought!

phpunit found that the fromRoman method in the Number class was not static and so not allowed to be called statically.

So, this pull request adds a lot of phpunit tests that found a bug that phpspec somehow missed and fixes a very tiny bug!

defrag commented:
Thanks for finding this bug, but rewritting all tests? Seems too much. I know that some of our specs arent really spec, but some integration tests for the fasades and we do have a plan to improve them, but peronally i wouldnt go that far and replace just all of them with phpunit. 
If you dont want to call static methods, why dont u just construct the objects behind them and call them non-staticly?

Zae commented:
That's what I'm doing right now, but I didn't know that was possible in PHP before writing all these tests.

I didn't throw away the specs, they are still there, they simply do the same right now.
I just thought I'd send the pull request, was a waste to simply throw them all away after doing such a good job in finding that bug :-P

norzechowicz commented:
@Zae nice! 
To be honest it was on my todo list for tomorrow. I was going to remove StringSpec, NumberSpec, DateTimeSpec and CollectionSpec, those specs are more interation tests rather than unit tests and I think it would be easier to test those facades with phpunit. 

Also according to PSR standards you should use 4 spaces instead of tab for indention as mentioned in [Contributing](https://github.com/coduo/php-humanizer/blob/master/CONTRIBUTING.md) file

norzechowicz commented on composer.json:
```diff
@@ -21,7 +21,9 @@
     "require-dev": {
-        "phpspec/phpspec": "2.0.*"
+        "phpspec/phpspec": "2.0.*",
+        "phpunit/phpunit": "^3.0|^4.0|^5.0",
```
Because this is only dev dependency I think we can use highest version ``^5.0``

norzechowicz commented on composer.json:
```diff
@@ -21,7 +21,9 @@
-        "phpspec/phpspec": "2.0.*"
+        "phpspec/phpspec": "2.0.*",
+        "phpunit/phpunit": "^3.0|^4.0|^5.0",
+        "mockery/mockery": "^0"
```
I would rather use prophecy that we already have than new mocking tool

norzechowicz commented on tests/unit/DateTime/DifferenceTest.php:
```diff
@@ -0,0 +1,48 @@
+<?php
+
+/**
```
Test classess also should have namespaces. 
Please take a look at https://github.com/isolate-org/unit-of-work/tree/master/tests folder. Just remember to specify this namespace in composer.json file just like here: https://github.com/isolate-org/unit-of-work/blob/master/composer.json#L18

norzechowicz commented on src/Coduo/PHPHumanizer/Number.php:
```diff
@@ -37,7 +37,7 @@ public static function toRoman($number)
     }
 
-    public function fromRoman($number)
+    public static function fromRoman($number)
```
Nice catch!

Zae commented:
Hey Norbert,
Glad you like it.

I didn't think about the spaces, my company uses most of PSR-2 only with tabs, I'll change it along with the other suggestions you made.

Do you want me to squash all the commits into one when i'm done?

Zae commented:
I did the changes and squashed all the test commits into one but left the one where I fixed the bug separate.

I added namespaces, changed to spaces

I changed the dev version requirement of phpunit, to 4.5 minimum, because that version includes prophecy for the first time and ^5.0 only supports >= php 5.6.

defrag commented on tests/unit/DateTimeTest.php:
```diff
@@ -0,0 +1,163 @@
+     */
+    public function preciseDifferenceDataProvider()
+    {
+        return array(
```
Maybe its a good idea to separate each test case per locale?

Zae replied:
I could, I thought it was easier and DRYer this way, simply add a new block to the dataprovider and you're good to go, the test code stays the same.

norzechowicz replied:
Im not totally sure if we should split those test cases. IMO this way is easier for new ppl to add tests for other languages.

norzechowicz commented on tests/unit/DateTime/DifferenceTest.php:
```diff
@@ -0,0 +1,52 @@
+     */
+    public function it_calculates_differences_correctly($expected, $date1, $date2, $instance, $is_past )
+    {
+        $diff = new \Coduo\PHPHumanizer\DateTime\Difference($date1, $date2);
```
Could you please move FQCN into use statement and use only ``$diff = new Difference($date1, $date2);`` here? 
Also please use only camel case in variable names. ``$isPast``

norzechowicz commented on tests/unit/DateTime/DifferenceTest.php:
```diff
@@ -0,0 +1,52 @@
+
+/**
+ * @author       Ezra Pool <ezra@tsdme.nl>
+ */
```
If its not a problem for you please remove those doc blocks. Not because I dont like it, we are just not using docblocks to mention authors, everything is in git log and after some time is hard to keep dockblocks and git log synchronized :P

Zae replied:
Sure no problem.

norzechowicz commented on tests/unit/DateTime/DifferenceTest.php:
```diff
@@ -0,0 +1,52 @@
+     * @param $instance
+     * @param $is_past
+     */
+    public function it_calculates_differences_correctly($expected, $date1, $date2, $instance, $is_past )
```
Could you please rename those methods in order to include "test" word at the beggining? 
For example: 
``public function test_calculating_differences_between_two_dates($expected, $firstDate, $secondDate, $instance, $isPast )``

norzechowicz commented on tests/unit/DateTime/DifferenceTest.php:
```diff
@@ -0,0 +1,52 @@
+            array(1, 	new \DateTime("2014-04-01"), 			new \DateTime("2014-03-01"), 			'Coduo\PHPHumanizer\DateTime\Unit\Month', 		true),
+            array(1, 	new \DateTime("2014-04-01"), 			new \DateTime("2014-05-01"), 			'Coduo\PHPHumanizer\DateTime\Unit\Month', 		false),
+            array(2, 	new \DateTime("2014-01-01"), 			new \DateTime("2012-01-01"), 			'Coduo\PHPHumanizer\DateTime\Unit\Year', 		true),
+            array(1, 	new \DateTime("2014-01-01"), 			new \DateTime("2015-01-01"), 			'Coduo\PHPHumanizer\DateTime\Unit\Year', 		false),
```
No need here for extran indentions

Zae replied:
Oh, I think that happened because I changed from tabs to spaces, remove the indent completely or simply fix it up so the columns align properly again?

norzechowicz replied:
I think u can remove them completely, just keep one space after comma

Zae commented:
Did the changes, but github says there are conflicts... Should I rebase on top of master?

norzechowicz commented:
Yes, please rebase against master

norzechowicz commented:
@Zae thanks! I gonna take it from here, we still need to add tests for languages that we currently support.

Zae commented:
rebase complete!

norzechowicz commented:
@Zae Im closing this PR in favor of #50, history stays untouched you are still main author of this change, thanks!.
